\documentclass[final, 8pt]{beamer}

\usepackage[size=custom,width=76.2,height=101.6,scale=0.85]{beamerposter}
\usetheme{confposter}

\newlength{\sepwid}
\newlength{\onecolwid}
\newlength{\twocolwid}
\newlength{\threecolwid}
\setlength{\paperwidth}{30in} % A0 width: 46.8in
\setlength{\paperheight}{40in} % A0 height: 33.1in
\setlength{\onecolwid}{0.3\paperwidth}
\setlength{\twocolwid}{0.464\paperwidth}
\setlength{\threecolwid}{0.708\paperwidth}
\setbeamersize{text margin left=0.05cm,text margin right=0.05cm}

\usepackage{graphicx}
\usepackage{booktabs}
\usepackage{natbib}
\input{../preamble.tex}

\title{Inference of Dynamic Regimes in the Microbiome}
\author{Kris Sankaran and Susan P. Holmes}
\institute{Department of Statistics, Stanford University}

\begin{document}

\begin{frame}

\begin{columns}
\begin{column}{\onecolwid}

\begin{block}{Abstract}
Many studies attempt to characterize microbiome stability and dynamics across
environments \citep{costello2012application, stein2013ecological,
  faust2015metagenomics}. These problems often have a flavor not just of time
series modeling but also of regime detection, the partitioning of time into
intervals with distinct behaviors, a problem with a rich history and application
to speech recognition \citep{fox2011sticky}, finance \citep{lee2009optimal}, EEG
analysis \citep{camilleri2014automatic} and geophysics
\citep{weatherley2002relationship}. In spite of the parallels, regime detection
methods are rarely used in microbiome data analysis.

We distill the core ideas of different regime detection methods, provide example
applications, and share reproducible code, making these techniques more
accessible to microbiome researchers. Ultimately, our goal is to describe types
of regime switching structure that, through careful modeling, can be
incorporated into studies of microbiome dynamics. Code for all examples here is
available at \url{https://github.com/krisrs1128/tsc\_microbiome}.
\end{block}

\begin{block}{Problem description}
Statistical analysis provides succinct representations of complex data. We can
think of the reduced representations as a type of data compression for human
interpretation, and as in any (lossy) compression, there is a choice of what
structure to preserve. Different reductions facilitate different comparisons --
for example, clustering bacteria allows easy comparison of taxonomic categories,
while clustering samples allows a comparison of full community states.

For regime detection, the comparisons we would like are
\begin{itemize}
\item For each species, can we assign time intervals to different dynamic
  regimes?
\item Can we define subsets of species which have similar patterns of behavior,
  in terms of these regimes?
\end{itemize}
\end{block}

\begin{block}{Case study background}
As a concrete use case for different regime detection methods, we reconsider the
data of \citep{dethlefsen2011incomplete}, which studied the response of
bacterial populations in the human gut to antibiotic treatments. Generally,
there is interest in characterizing dynamics and stability in the microbiome.

This study collected 16s sequencing measurements for three subjects sampled over
long time courses (~ 50 days), with two antibiotic interventions given in
between. Therefore, each subject served as their own control, and the effect of
antibiotic perturbations could be analyzed within the context of natural long
term variation.

There are strong inter-subject effects in this data, and we will focus on
subject F for further analysis. This subject had been found to have a microbiome
that only partially recovered after the first antibiotic time courses.
\end{block}

\begin{block}{Model-free analysis}
\begin{itemize}
\item Classification and Regression Trees [CART]: A tree that uses $X$ to
  predict $y$ provides a partition of the $X$ space in a way that ensures $y$
  has lower variation within than between partitions
  \citep{breiman1984classification}. We hierarchical clustering to obtain an
  ordering $s_{j} \in \{1, \dots, n_{\text{species}}\}$ across species, then
  model the count for the $j^{th}$ species in the $i^{th}$ sample as $y_{ij}
  \approx f\left(s_{j}, t_{i}\right)$ using $f \in \mathcal{T}$ the space of
  decision trees. The output is a fitted partition on $\left(s_{j},
  t_{i}\right)$, see Figure \ref{fig:rpart_binary}.

\begin{figure}[ht]
  \centering
  \includegraphics[width=\textwidth]{../figure/rpart_binary}
  \caption{CART partitions across species and timepoints. Each column here
    corresponds to an Amplicon Sequence Variant (ASV), and rows are timepoints.
    The three subjects (D, E, and F) are laid out side by side. Each rectangle
    in the figure represents the leaf node for a CART model fitted on
    presence-absence data, shaded by the fitted probability of being present in
    a given species by timepoint combination. The long horizontal beige bars in
    subjects D and F correspond to the antibiotics time courses. Note the
    incomplete recovery among some species in Subject F, corresponding to large
    beige partition rectangles.
    \label{fig:rpart_binary}}
\end{figure}
\end{itemize}
\end{block}
\end{column}

\begin{column}{\onecolwid}
\begin{itemize}
\item Hierarchical Clustering: 
We can visualize species abundances after ordering them according to a
hierarchical clustering tree. Note that clustering trees are invariant under
left-right swaps of branches, and we order branches so that the average
abundance of the left is always larger than those on the right. The resulting
ordered heatmap could potentially resolve partitions in the species by time
space. This does not explicitly account for temporal structure, but is simple to
both implement and interpret.

The figure generated by hierarchical clustering will be sensitive to several
choices,
\begin{itemize}
\item Transformations: Different transformations might be more effective
  representations of the underlying data.
\item Distance: Different distances are appropriate for different types of data.
\end{itemize}
\end{itemize}

\begin{block}{Parametric modeling}
\begin{itemize}
\item (Sticky) Hidden Markov Models (HMMs):

\begin{figure}[ht]
  \centering
  \includegraphics[width=0.8\textwidth]{../figure/hmm_mode}
  \caption{\label{fig:hmm_mode}}
\end{figure}


\item Switching Linear Dynamical Systems (SLDS):

\begin{figure}[ht]
  \centering
  \includegraphics[width=\textwidth]{../figure/slds_pca_scores}
  \caption{\label{fig:slds_pca_scores} }
\end{figure}

\item Bayesian Analysis of Simultaneous Changepoints (BASIC):
\begin{figure}[ht]
  \centering
  \includegraphics[width=\textwidth]{../figure/basic_bern_heatmap}
  \caption{\label{fig:basic_bern_heamtap} }
\end{figure}
\end{itemize}

\end{block}

\end{column}
\begin{column}{\onecolwid}

\begin{itemize}
\item Dynamic Tobit Models:
\begin{figure}[!p]
  \centering
  \includegraphics[width=0.8\textwidth]{../figure/abt_scan}
  \caption{\label{fig:abt_scan}}
\end{figure}
\end{itemize}

\begin{block}{Nonparametric modeling}
\begin{itemize}
\item Mixture of Gaussian Processes (mGPs)
\begin{figure}[!p]
  \centering
  \includegraphics[width=0.8\textwidth]{../figure/igp_abt_states}
  \caption{\label{fig:igp_abt_states}}
\end{figure}

\begin{figure}[!p]
  \centering
  \includegraphics[width=\textwidth]{../figure/igp_abt_fits}
  \caption{\label{fig:igp_abt_fits}}
\end{figure}

\item Sticky Hierarchical Dirichlet Process HMMs (HDP-HMMs)
\end{itemize}
\end{block}

 \bibliographystyle{plainnat}
 \bibliography{../bibliography.bib}
\end{column}
\end{columns}

\end{frame}
\end{document}
